webknossos-tracingstore/app/com/scalableminds/webknossos/tracingstore/tracings/volume/MergedVolume.scala

```diff
@@ -0,0 +1,135 @@
+
+import java.io.File
+
+import com.scalableminds.util.geometry.Point3D
+import com.scalableminds.util.tools.{ByteUtils, Fox}
+import com.scalableminds.webknossos.datastore.models.{BucketPosition, UnsignedInteger, UnsignedIntegerArray}
+import com.scalableminds.webknossos.datastore.services.DataConverter
+import com.scalableminds.webknossos.tracingstore.VolumeTracing.VolumeTracing.ElementClass
+import com.scalableminds.webknossos.tracingstore.tracings.ProtoGeometryImplicits
+import net.liftweb.common.Box
+
+import scala.collection.mutable
+import scala.concurrent.ExecutionContext
+
+class MergedVolume(elementClass: ElementClass, initialLargestSegmentId: Long = 0)
+    extends DataConverter
+    with ByteUtils
+    with VolumeDataZipHelper
+    with ProtoGeometryImplicits {
+  private var mergedVolume = mutable.HashMap.empty[BucketPosition, Array[UnsignedInteger]]
+  private var labelSets = mutable.ListBuffer[mutable.Set[UnsignedInteger]]()
+  private var labelMaps = mutable.ListBuffer[mutable.HashMap[UnsignedInteger, UnsignedInteger]]()
+  var largestSegmentId: UnsignedInteger = UnsignedInteger.zeroFromElementClass(elementClass)
+
+  def addLabelSetFromDataZip(zipFile: File)(implicit ec: ExecutionContext): Box[Unit] = {
+    val importLabelSet: mutable.Set[UnsignedInteger] = scala.collection.mutable.Set()
+    val unzipResult = withBucketsFromZip(zipFile) { (_, bytes) =>
+      val dataTyped =
+        UnsignedIntegerArray.fromByteArray(bytes, elementClass)
+      val nonZeroData = UnsignedIntegerArray.filterNonZero(dataTyped)
+      importLabelSet ++= nonZeroData
+    }
+    for {
+      _ <- unzipResult
+      _ = addLabelSet(importLabelSet)
+    } yield ()
+  }
+
+  def addLabelSetFromBucketStream(bucketStream: Iterator[(BucketPosition, Array[Byte])],
+                                  allowedResolutions: Set[Point3D]): Unit = {
+    val labelSet: mutable.Set[UnsignedInteger] = scala.collection.mutable.Set()
+    bucketStream.foreach {
+      case (bucketPosition, data) =>
+        if (allowedResolutions.contains(bucketPosition.resolution)) {
+          val dataTyped = UnsignedIntegerArray.fromByteArray(data, elementClass)
+          val nonZeroData: Array[UnsignedInteger] = UnsignedIntegerArray.filterNonZero(dataTyped)
+          labelSet ++= nonZeroData
+        }
+    }
+    addLabelSet(labelSet)
+  }
+
+  def addLabelSet(labelSet: mutable.Set[UnsignedInteger]): Unit = labelSets += labelSet
+
+  private def prepareLabelMaps(): Unit =
+    if (labelSets.isEmpty || (labelSets.length == 1 && initialLargestSegmentId == 0) || labelMaps.nonEmpty) {
+      ()
+    } else {
+      var i: UnsignedInteger = UnsignedInteger.zeroFromElementClass(elementClass)
+      if (initialLargestSegmentId > 0) {
+        labelMaps += mutable.HashMap.empty[UnsignedInteger, UnsignedInteger]
+        i = UnsignedInteger.fromLongWithElementClass(initialLargestSegmentId, elementClass)
+      }
+      labelSets.toList.foreach { labelSet =>
```

Comment:
The `toList` call is unnecessary in my opinion here because we don't rely on a return type here